File: backend/pkg/httpx/problem.go
// Package httpx: nền HTTP dùng chung cho các service Go — lỗi problem+json (RFC 9457), middleware,
// health check, chạy server với tắt êm (docs/09 §6, §12).
package httpx

import (
	"encoding/json"
	"errors"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"

	"github.com/NamHaiIT2HUST/DengueSense/backend/pkg/obsx"
)

const (
	problemTypeBase   = "https://denguesense.vn/errors/"
	problemContentTyp = "application/problem+json"
)

// FieldError mô tả lỗi validate trên một trường.
type FieldError struct {
	Field   string `json:"field"`
	Message string `json:"message"`
}

// Problem là thân phản hồi lỗi (khớp schema `Problem` trong contracts/openapi/public-v1.yaml).
type Problem struct {
	Type      string       `json:"type"`
	Title     string       `json:"title"`
	Status    int          `json:"status"`
	Code      string       `json:"code"`
	Detail    string       `json:"detail,omitempty"`
	Instance  string       `json:"instance,omitempty"`
	RequestID string       `json:"request_id"`
	Errors    []FieldError `json:"errors,omitempty"`
}

// Error là lỗi nghiệp vụ có mã ổn định (contracts/errors.md). Handler trả lỗi này, adapter HTTP
// chuyển thành problem+json ở MỘT chỗ (WriteProblem / ResponseError).
type Error struct {
	Status int
	Code   string
	Title  string
	Detail string
	Fields []FieldError
}

func (e *Error) Error() string { return e.Code + ": " + e.Title }

// New tạo lỗi mới. `code` PHẢI có trong contracts/errors.md.
func New(status int, code, title, detail string) *Error {
	return &Error{Status: status, Code: code, Title: title, Detail: detail}
}

// Các hàm dựng lỗi chuẩn (mã khớp contracts/errors.md).

func ValidationError(detail string, fields ...FieldError) *Error {
	e := New(http.StatusBadRequest, "common.validation_error", "Đầu vào không hợp lệ", detail)
	e.Fields = fields
	return e
}

func Unauthenticated(detail string) *Error {
	return New(http.StatusUnauthorized, "common.unauthenticated", "Chưa xác thực", detail)
}

func Forbidden(detail string) *Error {
	return New(http.StatusForbidden, "common.forbidden", "Không đủ quyền", detail)
}

func NotFound(detail string) *Error {
	return New(http.StatusNotFound, "common.not_found", "Không tìm thấy", detail)
}

func MethodNotAllowed() *Error {
	return New(http.StatusMethodNotAllowed, "common.method_not_allowed", "Phương thức không được hỗ trợ", "")
}

func NotImplemented() *Error {
	return New(http.StatusNotImplemented, "common.not_implemented", "Chức năng chưa được triển khai", "")
}

// Internal KHÔNG mang chi tiết nội bộ — chi tiết chỉ nằm trong log.
func Internal() *Error {
	return New(http.StatusInternalServerError, "common.internal_error", "Có lỗi xảy ra", "")
}

func DependencyUnavailable() *Error {
	return New(http.StatusServiceUnavailable, "common.dependency_unavailable", "Dịch vụ phụ thuộc không khả dụng", "")
}

// AsError chuyển mọi error thành *Error; lỗi lạ trở thành Internal (không lộ nội dung ra ngoài).
func AsError(err error) *Error {
	var e *Error
	if errors.As(err, &e) {
		return e
	}
	return Internal()
}

func (e *Error) problem(requestID, instance string) Problem {
	suffix := e.Code
	if i := strings.IndexByte(e.Code, '.'); i >= 0 {
		suffix = e.Code[i+1:]
	}
	return Problem{
		Type:      problemTypeBase + strings.ReplaceAll(suffix, "_", "-"),
		Title:     e.Title,
		Status:    e.Status,
		Code:      e.Code,
		Detail:    e.Detail,
		Instance:  instance,
		RequestID: requestID,
		Errors:    e.Fields,
	}
}

// WriteProblem ghi problem+json và dừng chuỗi handler.
func WriteProblem(c *gin.Context, e *Error) {
	body, err := json.Marshal(e.problem(obsx.RequestID(c.Request.Context()), c.Request.URL.Path))
	if err != nil { // không thể xảy ra với kiểu tĩnh này; giữ nhánh an toàn thay vì panic
		c.AbortWithStatus(http.StatusInternalServerError)
		return
	}
	c.Data(e.Status, problemContentTyp, body)
	c.Abort()
}

File: ai-service/app/serving/common/problem.py
"""Lỗi `application/problem+json` (RFC 9457) — cùng hình dạng và mã với `backend/pkg/httpx`.

Mã lỗi (`code`) PHẢI có trong `contracts/errors.md` (test kiểm). Lỗi không lường trước chỉ để lại chi tiết
trong log; thân phản hồi không bao giờ chứa stack trace, câu SQL, đường dẫn nội bộ hay giá trị đầu vào.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from fastapi import FastAPI, Request
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse
from starlette.exceptions import HTTPException as StarletteHTTPException

from app.serving.common.logging import request_id_var

PROBLEM_CONTENT_TYPE = "application/problem+json"
TYPE_BASE = "https://denguesense.vn/errors/"


@dataclass(frozen=True)
class FieldError:
    field: str
    message: str


class ApiError(Exception):
    """Lỗi nghiệp vụ có mã ổn định; handler `raise ApiError(...)`, adapter chuyển thành problem+json.

    Cố ý là lớp thường (không phải dataclass frozen): ngoại lệ frozen làm hỏng việc gán `__traceback__`.
    """

    def __init__(
        self,
        status: int,
        code: str,
        title: str,
        detail: str = "",
        fields: tuple[FieldError, ...] = (),
    ) -> None:
        super().__init__(f"{code}: {title}")
        self.status = status
        self.code = code
        self.title = title
        self.detail = detail
        self.fields = fields


def validation_error(detail: str, *fields: FieldError) -> ApiError:
    return ApiError(
        400, "common.validation_error", "Đầu vào không hợp lệ", detail, fields
    )


def not_found(detail: str = "") -> ApiError:
    return ApiError(404, "common.not_found", "Không tìm thấy", detail)


def method_not_allowed() -> ApiError:
    return ApiError(405, "common.method_not_allowed", "Phương thức không được hỗ trợ")


def internal_error() -> ApiError:
    return ApiError(500, "common.internal_error", "Có lỗi xảy ra")


def dependency_unavailable() -> ApiError:
    return ApiError(
        503, "common.dependency_unavailable", "Dịch vụ phụ thuộc không khả dụng"
    )


def problem_body(error: ApiError, instance: str) -> dict[str, Any]:
    suffix = error.code.split(".", 1)[-1].replace("_", "-")
    body: dict[str, Any] = {
        "type": TYPE_BASE + suffix,
        "title": error.title,
        "status": error.status,
        "code": error.code,
        "request_id": request_id_var.get() or "",
    }
    if error.detail:
        body["detail"] = error.detail
    if instance:
        body["instance"] = instance
    if error.fields:
        body["errors"] = [
            {"field": f.field, "message": f.message} for f in error.fields
        ]
    return body


def problem_response(error: ApiError, instance: str) -> JSONResponse:
    return JSONResponse(
        problem_body(error, instance),
        status_code=error.status,
        media_type=PROBLEM_CONTENT_TYPE,
    )


def _field_path(loc: tuple[Any, ...]) -> str:
    # loc = ("body", "origin_month") → "origin_month"; bỏ tiền tố nguồn (body/query/path/header).
    parts = [str(p) for p in loc]
    if parts and parts[0] in {"body", "query", "path", "header", "cookie"}:
        parts = parts[1:]
    return ".".join(parts) or "(root)"


def install_error_handlers(app: FastAPI) -> None:
    """Mọi đường lỗi của FastAPI/Starlette đều ra problem+json."""

    async def on_validation(request: Request, exc: Exception) -> JSONResponse:
        assert isinstance(exc, RequestValidationError)
        # KHÔNG dùng exc.errors()[i]["input"] / ["msg"] trực tiếp: có thể lặp lại giá trị người dùng gửi.
        fields = tuple(
            FieldError(_field_path(tuple(e["loc"])), _safe_message(str(e["type"])))
            for e in exc.errors()
        )
        return problem_response(
            validation_error("đầu vào không hợp lệ", *fields), request.url.path
        )

    async def on_http(request: Request, exc: Exception) -> JSONResponse:
        assert isinstance(exc, StarletteHTTPException)
        if exc.status_code == 404:
            err = not_found("đường dẫn không tồn tại")
        elif exc.status_code == 405:
            err = method_not_allowed()
        elif exc.status_code < 500:
            err = ApiError(
                exc.status_code, "common.validation_error", "Yêu cầu không hợp lệ"
            )
        else:
            err = internal_error()
        return problem_response(err, request.url.path)

    async def on_api_error(request: Request, exc: Exception) -> JSONResponse:
        assert isinstance(exc, ApiError)
        return problem_response(exc, request.url.path)

    app.add_exception_handler(RequestValidationError, on_validation)
    app.add_exception_handler(StarletteHTTPException, on_http)
    app.add_exception_handler(ApiError, on_api_error)


def _safe_message(error_type: str) -> str:
    """Thông điệp theo LOẠI lỗi pydantic (không chứa giá trị đầu vào)."""
    if error_type == "missing":
        return "trường bắt buộc"
    if error_type.endswith(("_type", "_parsing")):
        return "sai kiểu dữ liệu"
    if error_type.startswith("literal") or error_type == "enum":
        return "giá trị không thuộc tập cho phép"
    if error_type == "extra_forbidden":
        return "trường không được phép"
    return "giá trị không hợp lệ"
